### DISCUSSION TITLE
Heterogeneous graph classification

### DISCUSSION THREAD

**Discussion by Zytrus** [pyg-team/pytorch_geometric#4471] [Q&A] [answered]
Hello!

I was wondering how do you know, which model you're supposed to use for your current dataset?
My dataset consists of 3 types, let's call them A, B and C and 5 relation types. Now I want to make a graph classification, so I'm wondering out of the existing models and self made ones, when do you make one yourself and when do you just take an existing one?

Are there some kind of hints that can direct you to certain models? Like what kind of feature you have (in my case node degrees) or how large the dataset is?

When would it be smart to customize your own GNN? When is it more favorable to use an existing one?

Best regards, Zytrus

**Comment by rusty1s** [comment, marked as answer]
This is super hard to tell and is mostly based on experimental evaluation. In general, it is good to have a basic understanding of the dataset and which features might be important. This helps you to determine, e.g. the choice of aggregation function (mean, sum, max). For example, if you only have node degrees as features, you are solely able to learn from the graph structure, and so you would want to use layers specially designed for this task with the necessary expressive power (such as GINConv). Designing a new layer is certainly a more research centric approach. One usually starts with a baseline and tries to find out why it may underperform or how results can be improved.

**Reply by Zytrus** [reply]
Thank you for the answer! I've started testing out some existing GNNs for heterogenous graphs, but came upon the problem of a lot fo AttributeErrors... 
Say you made a dataset, do you have to define num_features, num_node_features, meta_data yourself?

Best regards, Zytrus

**Reply by rusty1s** [reply]
You can access the metadata via `dataset[0].metadata()`. For the number of input features across different node and edge types, we recommend that you make use of lazy initialization to simplify this process, i.e. by just passing `-1` as `in_channels`.

**Reply by Zytrus** [reply]
But what if I do need to access the num_features? When I go through the variables step by step, the num_classes is initialized, but num_features only show an AttributeError.... with no clear specification. I tried to follow back to the dictionary _store.. but still couldn't find the cause. Is it supposed to initialize itself?

**Reply by rusty1s** [reply]
The `num_features` call will fail for heterogeneous graphs since this may very across node types. As a simple workaround, you can receive it for each node type individually:
```python
for node_type in dataset[0].node_types:
    print(node_type, data[node_type].num_features)
```

**Reply by Zytrus** [reply]
Alright, thank you for the replies!

**Reply by rusty1s** [reply]
I integrated this as part of `HeteroData` in https://github.com/pyg-team/pytorch_geometric/pull/4504. You should now be able to do:
```python
print(hetero_data.num_node_features)
print(hetero_data.num_edge_features)
```
